core/src/test/scala/org/typelevel/paiges/PaigesScalacheckTest.scala

```diff
@@ -235,10 +209,6 @@ class PaigesScalacheckTest extends FunSuite {
     }
   }
 
-  test("Union invariant: `a.flatten == b.flatten`") {
```

Comment:
This test is no longer true.

Reply:
why isn't it:

https://github.com/typelevel/paiges/pull/117/files#diff-d14ba9b652696ca8a7646c84d04fc0e4R168

That looks to me like `Union(a, b)` implies that `a == b.flatten` and since flatten should be idempotent, `b.flatten.flatten == b.flatten` so, that implies `a.flatten == b.flatten` no?

Where am I going wrong there?

Reply:
`Union` is not only created by `grouped`, but also by `fill`, which is a lot
more complicated.  The law is indeed fulfilled by `grouped` (Checked by removing
`fill` from all generators.)  Assume x and y are flattenable, but z is unflattenable.  Then

    fill(sep, x :: y :: z :: Nil)

will be a `Union` but neither side will be flattenable, but will render differently.  For
a concrete example, consider

    fill(empty, List(empty, lineOrSpace, Line))

The lhs is 

    Concat(Empty, Concat(Empty, Concat(Text( ), Concat(Empty, Line))))

which simplifies to

    Concat(Text( ), Line)

The rhs to

    Concat(Empty, Concat(Empty, Concat(Union(Text( ), Union(Text( ), FlatAlt(Line, Text( )))), Concat(Empty, Line))))

which simplifies to

    Concat(Union(Text( ), Union(Text( ), FlatAlt(Line, Text( )))), Line)

While we'd like rhs.flatten to be the same as the lhs, it's not because of
the trailing Line, which makes it unflattenable.

Reply:
It just occurred to me that you may have meant that if there are any `Line` nodes, we could use the result of `flattenOption` or `flattenBoolean` with `None`.  (`flatten` returns the original element (perhaps normalized) if there is a `Line`).  If that's what you meant, than you are probably correct that the law holds, and I'd be happy to add it back as such.  It's just now a much weaker law, as I imagine most terms will have `Line` in some form.  I think there's a much stronger form of that law lurking somewhere, but I don't yet see an easy way to expose it.